ipa-core/src/helpers/transport/in_memory/config.rs

```diff
@@ -55,19 +53,28 @@ impl<F: Fn(&InspectContext, &mut Vec<u8>) + Send + Sync + 'static> StreamInterce
 
 /// The general context provided to stream inspectors.
 #[derive(Debug)]
-pub struct InspectContext {
-    /// The shard index of this instance.
-    /// This is `None` for non-sharded helpers.
-    pub shard_index: Option<ShardIndex>,
-    /// The MPC identity of this instance.
-    /// The combination (`shard_index`, `identity`)
-    /// uniquely identifies a single shard within
-    /// a multi-sharded MPC system.
-    pub identity: HelperIdentity,
-    /// Helper that will receive this stream.
-    pub dest: Cow<'static, str>,
-    /// Circuit gate this stream is tied to.
-    pub gate: Gate,
+pub enum InspectContext {
+    ShardMessage {
```

Comment:
I guess we don't need to intercept cross-shard messages unless we want to emulate network failures? Is there any other reason you want to corrupt them?

Reply:
I am considering three categories of messages:
1. Helper-to-helper messages in a non-sharded environment
2. Helper-to-helper messages in a sharded environment. These are similar to (1), but in order to uniquely identify a channel, the shard also needs to be specified. Step `Foo`, from shard 0 of H1 to shard 0 of H2, is a different channel from step `Foo`, from shard 1 of H1 to shard 1 of H2. This is the kind of message that I am adding support for in this PR.
3. Shard-to-shard messages in a sharded environment. i.e. step `Bar` from shard 0 of H1 to shard 1 of H1. I'm not intentionally targeting this case with these changes, although the situation may improve slightly. These messages aren't interesting to corrupt for the purpose of testing malicious security, because the shards of a helper fall within the same trust domain. Like you say, they might be interesting to corrupt to test a network failure, but that seems less important than testing the properties of malicious security protocols.

 `InspectContext::MpcMessage` covers cases (1) and (2), depending on whether `shard` is `None` or `Some(shard_index)`.

The reason I added `InspectContext::ShardMessage`, which is for case (3), even though it's not fully supported/tested, is to maintain roughly the same functionality of the code in `<Weak<InMemoryTransport<I>> as Transport>::send`. That code is used for all three categories of messages above. I also wanted to improve type safety for the `dest` member (previously `Cow<'static, str>`), to avoid potential confusion whether it should be interpreted as a helper ID or a shard index.

The `MaliciousHelper` utility is still restricted to cases (1) and (2). Since the goal of `MaliciousHelper` is to simplify writing tests that use the stream interceptor, it didn't seem appropriate to complicate the `MaliciousHelper` interface to try and support all three cases simultaneously. If we decide to write tests that fall under case (3), we can either code against the interceptor interface directly, or make a new `MaliciousShard` utility.